GTG/backends/backend_caldav.py

```diff
@@ -0,0 +1,831 @@
+            if tag in attached_tags:
+                return True
+
+        return False
+
+    def _create_todo(self, task):
+        tid = task.get_id()
+        uid = str(uuid.uuid4())
+
+        cal_name = self._calendar_name_for_task(task)
+        cal = self._find_calendar_by_name(cal_name)
+
+        # Add tag if there is a default
+        cal_name_tag = self._tag_to_gtg(cal_name)
+        if cal_name_tag not in [tag.get_name() for tag in task.get_tags()]:
+            task.add_tag(cal_name_tag)
+
+        return Todo(cal.add_todo(f"""BEGIN:VCALENDAR
+VERSION:2.0
+PRODID:-//Getting Things Gnome//CalDAV Backend//EN
+BEGIN:VTODO
+UID:{uid}
+END:VTODO
+END:VCALENDAR"""))
+
+    def _notify_user_about_default_calendar(self):
+        """ This function causes the infobar to show up with the message
+        about default calendar.
+        """
+        default_name = self._parameters['default-calendar-name']
+        message = _(
+            f"Could not find calendar {default_name}"
+            "Configure CalDAV to save in a calendar from this list : \n"
+        ) + self._calendar_names_info()
+        BackendSignals().interaction_requested(
+            self.get_id(), message,
+            BackendSignals().INTERACTION_INFORM, "on_continue_clicked")
+
+    def on_continue_clicked(self, *args):
+        """ Callback when the user clicks continue in the infobar
+        """
+        pass
+
+    def _calendar_names_info(self):
+        return ','.join(self._calendar_names())
+
+    def _id_to_tid(self, parent_url, uid, use_cache=True):
+        cal = self._find_calendar_matching_url(parent_url)
+        url = None
+        if use_cache:
+            try:
+                cache = self._cached_todos[parent_url]
+                url = cache['uids'][uid]
+            except KeyError:
+                pass
+        if url is None:
+            todo = cal.todo_by_uid(uid) # network lookup
+            url = str(todo.url)
+        return self.sync_engine_url.get_local_id(url)
+
+    def _tid_to_id(self, tid, use_cache=True):
+        try:
+            url = self.sync_engine_url.get_remote_id(tid)
+        except KeyError:
+            return None, None
+        parent_url = None
+        uid = None
+        if use_cache:
+            for cal_url in self._cached_todos:
+                if url.startswith(cal_url):
+                    parent_url = cal_url
+                    try:
+                        uid = self._cached_todos[cal_url]['urls'][url]
+                    except KeyError:
+                        pass
+                    break
+        if parent_url is None or uid is None:
+            ev = self._find_event_by_url(url) # network lookup
+            try:
+                parent_url = ev.parent.url
+                uid = ev.instance.vtodo.uid.value
+            except AttributeError:
+                parent_url = None
+                uid = None
+        return parent_url, uid
+
+class Todo():
```

Comment:
I know it's obvious but a docstring here to explain what the class represents, technical details, etc. would be nice